src/Program.ts

```diff
@@ -277,6 +280,27 @@ export class Program {
         return this.getComponent(componentName)?.scope;
     }
 
+    /**
+     * Get the type definitions for the specified file.
+     * The results are cached for future performance boosts.
+     */
+    private async getTypeDefinitionsForFile(pathAbsolute: string) {
+        //you can only get type definitions for .brs files
+        if (!pathAbsolute.endsWith('.brs')) {
+            return;
+        }
+
+        let typeDefinitionPath = pathAbsolute.replace(/.brs$/, '.d.bs');
```

Comment:
Do you really need to do this lookup? When are the definitions concretely needed?
`.d.bs` files will be collected like other files in the Program so they will be in the files map.

`BrsFile`s could have a `get definitions` to dynamically look up for a definition file (collcted in Program)

Reply:
Currently within this PR in the `BrsFile.ts`, if we detect that there's a definition file for the `BrsFile`, we bypass parsing the `.brs` file. This should significantly reduce parse times for libraries that include `d.bs` files since we don't need to parse the whole `.brs` file, only its definitions. 

The main problem is, we aren't guaranteed _when_ the files will be passed into `Program`. There are three methods to adding files to the program, all of which will emit files when they find them, in no apparent order

 - The vscode file watcher while in language server mode (of which we have no control other than file matcher globs)
 - The glob search during a single cli run
 - the chokidar file watcher during a `--watch` cli run

This may lead to race conditions where the `.brs` files are handed to the program _before_ the `.d.bs` file.

If we don't detect the `d.bs` file early, then we will end up paying parse costs for every library file, only to ignore most of what we parsed once we detect the definitions file in your proposed `get definitions` getter.

Reply:
Ok if you want to avoid parsing it makes some sense.

Instead of doing this `getFileContents` though what about: 
if `.d.ts` file exits, and it's not in the file map, then automatically add and parse it, and link its `BrsFile` to the `.brs`'s `BrsFile`.

Also you may still want to open the library's `.brs` file in VSCode, then you still want the editor to understand the file. You will need then to parse the real file.

Reply:
>Instead of doing this getFileContents (and parse the .d.bs twice) 

In the current logic, the `d.bs` file will only be parsed once (within its corresponding `.brs` file). Also, since most developers aren't actively editing `.brs` files that have associated `d.bs` files, the current design won't cause performance issues for normal development. 

>if .d.bs file exits, and it's not in the file map, then automatically add and parse it, and link its BrsFile to the .brs's BrsFile.

I like this plan and want to move that direction eventually. However, I think this will need to be a future enhancement. I am trying very hard to get basic type definition support sooner rather than later to correspond with the ropm release and the upcoming meetup and roku developer "converence". 

> Also you may still want to open the library's .brs file in VSCode, then you still want the editor to understand the file. You will need then to parse the real file.

Very valid use case, and something we definitely need to support at some point. In general, most developers will not be hand-writing `d.bs` files in the short-term (within the next 1-3 months). Especially not with `d.bs` files being brand new. So I think we'll address this use case at a later date. 

Most of the implementation details of `d.bs` files are internal, so we can change their functionality as the design evolves.

Reply:
I mean, aren't the `.d.bs` files part of the files you will collect anyway?

Reply:
If you add a library with typedefs in your project, brighterscript will collect all the brs and the bs files, which will include the d.bs one's.

Reply:
resolving this because it has been implemented.